**One-bin  2A <-> B reaction, COMPARING 1st-order and 2nd-order kinetics in forward direction;
reverse direction 1-st order**

LAST REVISED: June 13, 2022

In [1]:
import set_path
set_path.add_ancestor_dir_to_syspath(3)  # The number of levels to go up 
                                         # to reach the project's home from the folder containing this notebook

Added 'D:\Docs\- MY CODE\BioSimulations\life123-Win7' to sys.path


In [2]:
from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log

In [3]:
# Initialize the system
chem_data = chem(diffusion_rates=[0.1, 0.1], names=["A", "B"])  # NOTE: diffusion_rates not used for now

rxn = Reactions(chem_data)

# Reaction  2A <-> B , FOR NOW with 1st-order kinetics in both directions
rxn.add_reaction(reactants=[(2, "A")], products=["B"], forward_rate=5., reverse_rate=2.)

bio.initialize_system(n_bins=1, chem_data=chem_data, reactions=rxn)

bio.set_all_uniform_concentrations( [3., 5.] )

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [3.]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.]


In [4]:
rxn.describe_reactions()

Number of reactions:  1
0: 2 A <-> B  (Rf = 5.0 / Rb = 2.0)


In [5]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(2, 0, 1)] <-> [(1, 1, 1)]   ; Fwd: 5.0 / Back: 2.0


# FOR NOW with 1st-order kinetics in both directions

In [6]:
# First step
bio.react(time_step=0.02, n_steps=1)
bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [2.8]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.1]


In [7]:
"""
1 bins and 2 species:
 [[2.8]
 [5.1]]
"""


# Numerous more steps
bio.react(time_step=0.02, n_steps=20)

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [2.16928427]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.41535786]


Consistent with the 5/2 ratio of forward/reverse rates (and the 1st order reactions),
the systems settles in the following equilibrium:  
[A] = 2.16928427 , [B] = 5.41535786

In [8]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
print(f"Ratio of equilibrium concentrations: {B_eq / A_eq}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations: 2.496379996594894
Ratio of forward/reverse rates: 2.5


# STARTING OVER, this time with 2nd-order kinetics in the forward reaction

In [9]:
rxn.clear_reactions()

In [10]:
# Reaction  2A <-> B , NOW WITH 2nd-order kinetics in the forward direction
rxn.add_reaction(reactants=[(2, "A", 2)], products=["B"], forward_rate=5., reverse_rate=2.)

bio.set_all_uniform_concentrations( [3., 5.] )

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [3.]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.]


In [11]:
rxn.describe_reactions()

Number of reactions:  1
0: 2 A <-> B  (Rf = 5.0 / Rb = 2.0) | 2-th order in reactant A


In [12]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(2, 0, 2)] <-> [(1, 1, 1)]   ; Fwd: 5.0 / Back: 2.0


In [13]:
# First step
bio.react(time_step=0.02, n_steps=1)
bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [1.6]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.7]


In [14]:
"""
1 bins and 2 species:
 [[1.6]
 [5.7]]
"""

# Numerous more steps
bio.react(time_step=0.02, n_steps=20)

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [1.51554944]
  Species 1 (B). Diff rate: 0.1. Conc:  [5.74222528]


In [15]:
"""
1 bins and 2 species:
 [[1.51554944]
 [5.74222528]]
"""

A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
print(f"Ratio of equilibrium concentrations ((B_eq) / (A_eq **2)) : {(B_eq) / (A_eq **2)}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations ((B_eq) / (A_eq **2)) : 2.4999999997397597
Ratio of forward/reverse rates: 2.5
